File: csharp/src/SolverRegistry.cs
namespace AdventOfCode;

public static class SolverRegistry
{
    private static readonly Dictionary<(int Year, int Day), AdventDay> _solvers;

    static SolverRegistry()
    {
        _solvers = typeof(Program)
            .Assembly
            .GetTypes()
            .Where(t => !t.IsAbstract && typeof(AdventDay).IsAssignableFrom(t))
            .Select(t => (AdventDay)Activator.CreateInstance(t)!)
            .ToDictionary(s => (s.Year, s.Day));
    }

    public static AdventDay Get(int year, int day) =>
        _solvers.TryGetValue((year, day), out var solver)
            ? solver
            : throw new InvalidOperationException($"No solver for {year} day {day}");
}

File: fsharp/src/SolverRegistry.fs
namespace AdventOfCode

open System

module SolverRegistry =

    let private solvers : Map<(int * int), IAdventDay> =
        AppDomain.CurrentDomain.GetAssemblies()
        |> Seq.collect (fun asm -> asm.GetTypes())
        |> Seq.filter (fun t ->
            not t.IsAbstract &&
            typeof<IAdventDay>.IsAssignableFrom t)
        |> Seq.choose (fun t ->
            try
                Some (Activator.CreateInstance(t) :?> IAdventDay)
            with _ ->
                None)
        |> Seq.map (fun s -> ((s.Year, s.Day), s))
        |> Map.ofSeq

    let get year day =
        match solvers |> Map.tryFind (year, day) with
        | Some solver -> solver
        | None ->
            invalidOp $"No solver registered for {year} day {day}"
